controller/instance_manager_controller.go

```diff
@@ -247,9 +255,101 @@ func (imc *InstanceManagerController) syncInstanceManager(key string) (err error
 		return imc.ds.RemoveFinalizerForInstanceManager(im)
 	}
 
+	existingIM := im.DeepCopy()
+	defer func() {
+		if err == nil && !reflect.DeepEqual(existingIM, im) {
+			_, err = imc.ds.UpdateInstanceManager(im)
+		}
+		if apierrors.IsConflict(errors.Cause(err)) {
+			logrus.Debugf("Requeue %v due to conflict", key)
+			imc.enqueueInstanceManager(im)
+			err = nil
+		}
+	}()
+
+	image, err := imc.ds.GetEngineImage(im.Spec.EngineImage)
+	if err != nil {
+		if datastore.ErrorIsNotFound(err) {
+			logrus.Infof("Engine image %v for instance manager %v has been deleted", im.Spec.EngineImage, key)
+			return nil
+		}
+		return err
+	}
+
+	pod, err := imc.ds.GetInstanceManagerPod(im.Name)
+	if err != nil {
+		return errors.Wrapf(err, "cannot get pod for instance manager %v", im.Name)
+	}
+
+	// We're handling the Instance Manager from a Node that isn't supposed to be responsible for it. This could mean
+	// that the Node it belongs to went down, or its newly scheduled.
+	if im.Spec.NodeID != imc.controllerID {
+		im.Status.CurrentState = types.InstanceManagerStateUnknown
+		return nil
+	}
+
+	if im.Status.CurrentState == types.InstanceManagerStateError {
+		if err := imc.cleanupInstanceManager(im); err != nil {
+			return err
+		}
+		if err := imc.createInstanceManagerPod(im, image); err != nil {
+			return err
+		}
+		im.Status.CurrentState = types.InstanceManagerStateStarting
+		return nil
+	}
+
+	if pod == nil {
+		if im.Status.CurrentState != types.InstanceManagerStateStopped {
+			im.Status.CurrentState = types.InstanceManagerStateError
+			return nil
+		}
+		if err := imc.createInstanceManagerPod(im, image); err != nil {
+			return err
+		}
+		im.Status.CurrentState = types.InstanceManagerStateStarting
+		return nil
+	}
+
+	switch pod.Status.Phase {
+	case v1.PodPending:
+		if im.Status.CurrentState == types.InstanceManagerStateUnknown {
+			im.Status.CurrentState = types.InstanceManagerStateStarting
+		} else if im.Status.CurrentState != types.InstanceManagerStateStarting {
+			im.Status.CurrentState = types.InstanceManagerStateError
```

Comment:
When did you see this happen? Pending should be the phase before running.

Reply:
I actually haven't witnessed this happening during normal behavior. This was put in when Shuo and I were discussing the design of the controller, as a fail-safe in case some unexpected behavior occurred or something else was interfering with our `Pods`.

Reply:
OK, we can set it to error but please make sure log the reason. I tend to think it's unnecessary but we can check the log later.